packages/block/template.ts

```diff
@@ -0,0 +1,123 @@
+import { AbstractBlock, EditType, Hole } from './types';
+import type { Edit, VElement } from './types';
+
+const IS_VOID_ELEMENT =
+  /area|base|basefont|bgsound|br|col|command|embed|frame|hr|image|img|input|isindex|keygen|link|menuitem|meta|nextid|param|source|track|wbr/i;
+
+export const renderToTemplate = (
+  vnode: VElement,
+  edits: Edit[] = [],
+  path: number[] = [],
+): string => {
+  let props = '';
+  let children = '';
+  const current: Edit = {
+    path, // The location of the edit in in the virtual node tree
+    edits: [], // Occur on mount + patch
+    inits: [], // Occur before mount
+  };
+
+  for (let name in vnode.props) {
+    const value = vnode.props[name];
+    if (name === 'key' || name === 'ref' || name === 'children') {
+      continue;
+    }
+    if (name === 'className') name = 'class';
+    if (name === 'for') name = 'htmlFor';
+
+    if (name.startsWith('on')) {
+      const isValueHole = value instanceof Hole;
+      // Make objects monomorphic
+      current.edits.push({
+        type: EditType.Event,
+        listener: isValueHole ? value.key : value,
+        name,
+        hole: isValueHole ? value.key : undefined,
+        value: undefined,
+        index: undefined,
+        patch: undefined,
+        block: undefined,
+      });
+
+      continue;
+    }
+
+    if (value instanceof Hole) {
+      current.edits.push({
+        type: EditType.Attribute,
+        hole: value.key,
+        name,
+        listener: undefined,
+        value: undefined,
+        index: undefined,
+        patch: undefined,
+        block: undefined,
+      });
+      continue;
+    }
+
+    if (value) props += ` ${name}="${String(value)}"`;
+  }
+
+  if (IS_VOID_ELEMENT.test(vnode.tag)) return `<${vnode.tag}${props} />`;
```

Comment:
I'm not sure if a regex is faster than e.g using an array of tags (`TAGS.includes(vnode.tag)`)

Reply:
Seems like you're right https://perf.link/#eyJpZCI6ImxmNWNnZzRjZDkxIiwidGl0bGUiOiJGaW5kaW5nIG51bWJlcnMgaW4gYW4gYXJyYXkgb2YgMTAwMCIsImJlZm9yZSI6ImNvbnN0IElTX1ZPSURfRUxFTUVOVCA9IC9hcmVhfGJhc2V8YmFzZWZvbnR8Ymdzb3VuZHxicnxjb2x8Y29tbWFuZHxlbWJlZHxmcmFtZXxocnxpbWFnZXxpbWd8aW5wdXR8aXNpbmRleHxrZXlnZW58bGlua3xtZW51aXRlbXxtZXRhfG5leHRpZHxwYXJhbXxzb3VyY2V8dHJhY2t8d2JyL2k7XG4gIFxuY29uc3Qgdm9pZEVsZW1lbnRzID0gW1xuICAgIFwiYXJlYVwiLFxuICAgIFwiYmFzZVwiLFxuICAgIFwiYmFzZWZvbnRcIixcbiAgICBcImJnc291bmRcIixcbiAgICBcImJyXCIsXG4gICAgXCJjb2xcIixcbiAgICBcImNvbW1hbmRcIixcbiAgICBcImVtYmVkXCIsXG4gICAgXCJmcmFtZVwiLFxuICAgIFwiaHJcIixcbiAgICBcImltYWdlXCIsXG4gICAgXCJpbWdcIixcbiAgICBcImlucHV0XCIsXG4gICAgXCJpc2luZGV4XCIsXG4gICAgXCJrZXlnZW5cIixcbiAgICBcImxpbmtcIixcbiAgICBcIm1lbnVpdGVtXCIsXG4gICAgXCJtZXRhXCIsXG4gICAgXCJuZXh0aWRcIixcbiAgICBcInBhcmFtXCIsXG4gICAgXCJzb3VyY2VcIixcbiAgICBcInRyYWNrXCIsXG4gICAgXCJ3YnJcIlxuXSIsInRlc3RzIjpbeyJuYW1lIjoiRmluZCBpdGVtIDEwMCIsImNvZGUiOiJ2b2lkRWxlbWVudHMuaW5jbHVkZXMoJ2xpbmsnKSIsInJ1bnMiOls4MTIwMDAsMjAzMzAwMCwxMDc5MDAwLDEyMDEwMDAsMTU4OTAwMCwzMzEwMDAwLDMzNjAwMDAsMjgzNzAwMCwxOTIxMDAwLDExMjQwMDAsNzA0MDAwLDc3MDAwLDEyMjYwMDAsOTEwMDAsNTUwMDAsMTYzNjAwMCwxNDAwMDAwLDEyNDAwMCwxMDM1MDAwLDMxMDkwMDAsMjU5ODAwMCw4NjIwMDAsMTYzNjAwMCwxMzExMDAwLDIwMjAwMDAsMTc0MjAwMCwzMjAwMDAsMTE5NjAwMCwxMDA3MDAwLDE2ODgwMDAsMTQxMDAwLDE2MTYwMDAsNjAwMCwxOTU4MDAwLDc3ODAwMCw3MDYwMDAsMjc4MDAwLDE1MTcwMDAsODI2MDAwLDE3NjAwMDAsMTAwMCwxNzQwMDAsMjExMDAwLDE2MzkwMDAsMjIwMzAwMCw5NTcwMDAsMjI5MDAwLDI0MTAwMDAsODQ2MDAwLDYzMTAwMCw5MjAwMDAsMTA3MDAwLDM0NTQwMDAsMjQ1MzAwMCw2OTIwMDAsMjA1NjAwMCwyNzkzMDAwLDE1MzYwMDAsMTcwNzAwMCwxNzMwMDAsMTEwMDAwMCwxMjMxMDAwLDIwMDAwLDc3OTAwMCwyMDEwMDAsMTkzNjAwMCwyMTY0MDAwLDczMTAwMCw2NDIwMDAsMjAwMDAsMzIzNjAwMCwzNDUwMDAsMjM1MTAwMCwyMDM1MDAwLDI1MjAwMCwxMTQzMDAwLDc4OTAwMCw1MTYwMDAsMjYzMjAwMCwxMTY3MDAwLDk5NzAwMCwyMjEwMDAsNzQ2MDAwLDE3NDQwMDAsMTkwMDAsNjQyMDAwLDE5MDAwLDM4OTAwMCwxMDU3MDAwLDI3MjAwMCwyNzIwMDAwLDE4MTMwMDAsMTc1OTAwMCw0MjQwMDAsMTQ1NDAwMCw0OTUwMDAsMTI1MzAwMCw2MTAwMDAsMzIwMDAwLDI0OTQwMDBdLCJvcHMiOjEyMDYxOTB9LHsibmFtZSI6IkZpbmQgaXRlbSAyMDAiLCJjb2RlIjoiSVNfVk9JRF9FTEVNRU5ULnRlc3QoJ2xpbmsnKSIsInJ1bnMiOlsxMDAwLDIwMDAsMTAwMCwyMDkxMDAwLDQ4MzAwMCwxMDAwLDExMDcwMDAsMTI4MDAwMCw2OTkwMDAsMTAwMCwxMDAwLDYyNjAwMCw1NDAwMCwyMDAwLDEwMDAsNzQxMDAwLDE3MTAwMCwyMDAwLDE4MTkwMDAsMTAwMCw3MzAwMDAsMjAwMCwxMDAwLDYwNzAwMCw2NjcwMDAsMTAwMCw3NjEwMDAsMjEzMDAwLDIwMDAsMTAwMCw1MDEwMDAsMjAwMCw2NDAwMDAsNTEwMDAwLDg5NjAwMCw1NjMwMDAsNTg3MDAwLDEwMDAsMjAwMCwyMDAwLDEwMDAsMTAwMCwyNjkwMDAsMTQ1MDAwLDQ5MjAwMCwyMDAwMCwxNTE2MDAwLDEwMDAsMTIyMDAwLDYxNzAwMCwyMDAwLDEwMDAsMjEyOTAwMCw3NjMwMDAsMjAwMCwxNDE4MDAwLDMzODAwMCw0ODYwMDAsMTAwMCwzNjMwMDAsNjkxMDAwLDEyODAwMCwxMDAwLDIwMDAsMjAwMCwxMzUyMDAwLDIwMDAsNDI2MDAwLDI3ODAwMCwxMDgwMDAwLDU4MDAwLDEwMDAsMTYyNjAwMCwxMDAwLDEwMDAsMTE3NTAwMCwxOTAwMDAsMjAwMCw1ODYwMDAsMjAwMCwyMDAwLDIwMDAsMTkwMDAsMTAwMCwyMDAwLDIwMDAsMTAwMCwxOTAwMCwyMDAwLDIzMzAwMCwxMDAwLDI3ODAwMCwyMDAwLDEwMDAsMTc2MDAwLDEwMDAsMTAwMCwyMDAwLDEwMDAsMjAwMF0sIm9wcyI6MzI4MTIwfV0sInVwZGF0ZWQiOiIyMDIzLTAyLTE3VDA0OjEzOjEwLjc4MloifQ%3D%3D

Reply:
But [`Set.prorotype.has` should be faster than `Array.prototype.includes`].

Reply:
Implemented with Set